lib/commands/context.js

```diff
@@ -0,0 +1,472 @@
+    }
+  } else {
+    if (this.isRealDevice()) {
+      this.remote = new WebKitRemoteDebugger();
+      return this.remote.pageArrayFromJson();
+    } else {
+      this.remote = new RemoteDebugger({
+        bundleId: this.caps.bundleId,
+        useNewSafari: this.useNewSafari(),
+        pageLoadMs: this.pageLoadMs,
+        platformVersion: this.caps.platformVersion
+      });
+
+      let appInfo = await this.remote.connect();
+      if (!appInfo) {
+        logger.debug('Unable to connect to the remote debugger.');
+        this.remote = null;
+        return [];
+      }
+      pageArray = await this.remote.selectApp(this.caps.webviewConnectRetries);
+      this.remote.on(RemoteDebugger.EVENT_PAGE_CHANGE, this.onPageChange.bind(this));
+
+      // var loopCloseRuns = 0;
+      // var loopClose = function () {
+      //   loopCloseRuns++;
+      //   if (!isDone && loopCloseRuns < 3) {
+      //     this.closeAlertBeforeTest(function (err, didDismiss) {
+      //       if (!didDismiss) {
+      //         setTimeout(loopClose, 1000);
+      //       }
+      //     });
+      //   }
+      // }.bind(this);
+      // setTimeout(loopClose, 4000);
+      await B.delay(4000);
+      // TODO: figure out how to do the alert stuff
+    }
+  }
+
+  if (pageArray.length === 0) {
+    // we have no web frames, so disconnect from the remote debugger
+    this.stopRemote();
+  }
+  return pageArray;
+};
+
+extensions.onPageChange = async function (pageArray) {
+  logger.debug(`Remote debugger notified us of a new page listing: ${JSON.stringify(pageArray)}`);
+  if (this.selectingNewPage) {
+    logger.debug('We are in the middle of selecting a page, ignoring');
+    return;
+  }
+  let newIds = [];
+  let newPages = [];
+  let keyId = null;
+  for (let page of pageArray) {
+    let id = page.id.toString();
+    newIds.push(id);
+    if (page.isKey) {
+      keyId = id;
+    }
+    if (!_.contains(this.contexts, id)) {
+      newPages.push(id);
+      this.contexts.push(id);
+    }
+  }
+
+  let newPage = null;
+  if (this.curContext === null) {
+    logger.debug('We do not appear to have window set yet, ignoring');
+  } else if (newPages.length) {
+    logger.debug(`We have new pages, going to select page '${newPages[0]}'`);
+    newPage = newPages[0];
+  } else if (!_.contains(newIds, this.curContext.toString())) {
+    logger.debug('New page listing from remote debugger does not contain ' +
+                 'current window; assuming it is closed');
+    if (keyId !== null) {
+      logger.debug(`Debugger already selected page '${keyId}', ` +
+                   `confirming that choice.`);
+    } else {
+      logger.error('Do not have our current window anymore, and there ' +
+                   'are not any more to load! Doing nothing...');
+      return;
+    }
+    this.curContext = keyId;
+    // this.remote.pageIdKey = parseInt(keyId, 10);
+    newPage = keyId;
+  } else {
+    let dirty = () => {
```

Comment:
if we just call this one time, let's make it a self-executing anonymous function so that `dirty` is just a bool. this also deserves a comment.